### PR TITLE
Add a hotkey to cycle harvesters

### LINKED ISSUE #17453: Add a hotkey to cycle harvesters
In CNC games it is crucial to protect your harvesters and to not have idling harvesters. A hotkey to cycle through all harvester could help with identifying idle ones. 

CNC3 has a hotkey called "Next harvester" that does just that. I've not played that game much beyond the single player campaign so I can't tell how useful such a hotkey really is but in theory it sound reasonable. Players with experience in CNC3 can share their opinion.

Another way to go about it is to have the hotkey cycle only idling harvesters. This is how it works in Age of Empires with villagers.

### PR DESCRIPTION
**Pull request by dragunoff**
A couple of remarks:
- Bound to <kbd>N</kbd> to match the default in CNC3 but I'd be fine with making it undefined
- I thought about putting this in Mods.Cnc because harvesters are specific to CNC games

Closes #17453

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+			var harvesters = world.ActorsHavingTrait<Harvester>()
+				.Where(a => a.Owner == player)
+				.ToList();
+
+			if (!harvesters.Any())
+				return true;
+
+			var next = harvesters
+				.SkipWhile(b => !selection.Contains(b))
+				.Skip(1)
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs]: Why are you doing this?

**Comment by Mailaender** [comment]
This needs an out of map bounds check for Dune 2000 during the harvester carryall delivery. Otherwise works as promised. :+1:

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs (hunk 1 above)
Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs]: I have not given this much thought - I copied the structure and logic from `CycleProductionActorsHotkeyLogic.cs`. If there's a better way to do it I'm open to suggestions on how to improve it.

**Comment by dragunoff** [comment]
> This needs an out of map bounds check for Dune 2000 during the harvester carryall delivery.

@Mailaender Could you give me a pointer on how to do that? An example in the code where a similar thing is done would be great. I tried to find but couldn't.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs]: 2020.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+			selection = world.Selection;
+			this.world = world;
+		}
+
+		protected override bool OnHotkeyActivated(KeyInput e)
+		{
+			var player = world.RenderPlayer ?? world.LocalPlayer;
+
+			var harvesters = world.ActorsHavingTrait<Harvester>()
+				.Where(a => a.Owner == player)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs]: Adding a check for `a.IsInWorld` here will probably fix the issue with carryalls.

**Comment by dragunoff** [comment]
Rebased and fixed the issue with harvesters in transports and caryalls.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs (hunk 2 above)
Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs]: Updated.

**Review by Mailaender — APPROVED** [review]

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/31f7a03fbf265455f8988f96b7ffbdd4afa74e7e...99af84d1aca0059ba9eef930ecece4d4205367b5)

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs (hunk 1 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleHarvestersHotkeyLogic.cs]: If we don't skip one more here we'll always end up with the currently selected one.

File: mods/common/hotkeys/game.yaml [hunk 4]
```diff
@@ -18,6 +18,10 @@ SelectUnitsByType: W
 	Description: Select units by type
 	Types: World, Player, Spectator
 
+CycleHarvesters: N
```
Comment (pchote) [inline mods/common/hotkeys/game.yaml]: I bet that putting this right next to the dangerously confusing `M` key on the keyboard is going to cause trouble. I'd rather we fix this by adding a visual indicator for mute than the change the hotkey.

Comment (dragunoff) [reply mods/common/hotkeys/game.yaml]: Agreed. I often press the mute key by accident. Filed #17549

**Pull request merged**